app/code/Magento/Inventory/Test/Unit/Model/ReservationBuilderTest.php

```diff
@@ -0,0 +1,111 @@
+use Magento\InventoryApi\Api\Data\ReservationInterface;
+
+class ReservationBuilderTest extends \PHPUnit\Framework\TestCase
+{
+    /** @var  ObjectManager */
+    private $objectManager;
+
+    /** @var ReservationBuilderValidatorInterface */
+    private $reservationBuilderValidator;
+
+    /** @var  ReservationInterface */
+    private $reservation;
+
+    /** @var  ValidationResult */
+    private $validationResult;
+
+    /** @var StringService */
+    private $stringService;
+
+    protected function setUp()
+    {
+        $this->objectManager = $this->getMockBuilder(ObjectManagerInterface::class)->getMock();
+        $this->reservationBuilderValidator = $this->getMockBuilder(ReservationBuilderValidatorInterface::class)->getMock();
+        $this->stringService = $this->getMockBuilder(StringService::class)->getMock();
+        $this->reservation = $this->getMockBuilder(ReservationInterface::class)->getMock();
+        $this->validationResult = $this->getMockBuilder(ValidationResult::class)->disableOriginalConstructor()->getMock();
+    }
+
+    private function createReservationBuilder(): ReservationBuilder
+    {
+        return (new ObjectManager($this))->getObject(
+            ReservationBuilder::class,
+            [
+                'objectManager' => $this->objectManager,
+                'reservationBuilderValidator' => $this->reservationBuilderValidator,
+                'stringService' => $this->stringService,
+            ]
+        );
+    }
+
+    /**
+     * @dataProvider getValidReservationData
+     */
+    public function testBuildValidReservation($reservationData, $reservationMappedData)
+    {
+        $this->stringService
+            ->expects($this->once())
+            ->method('convertArrayKeysFromSnakeToCamelCase')
+            ->with($reservationData)
+            ->willReturn($reservationMappedData);
+
+        $this->objectManager
+            ->expects($this->once())
+            ->method('create')
+            ->with(ReservationInterface::class, $reservationMappedData)
+            ->willReturn($this->reservation);
+
+        $this->validationResult
+            ->expects($this->once())
+            ->method('isValid')
+            ->willReturn(true);
+
+        $this->reservationBuilderValidator
+            ->expects($this->once())
+            ->method('validate')
+            ->willReturn($this->validationResult);
+
+        /** @var ReservationBuilderInterface $reservationBuilder */
+        $reservationBuilder = $this->createReservationBuilder();
+        $reservationBuilder->setStockId($reservationData[ReservationInterface::STOCK_ID]);
+        $reservationBuilder->setSku($reservationData[ReservationInterface::SKU]);
+        $reservationBuilder->setQuantity($reservationData[ReservationInterface::QUANTITY]);
+        $reservationBuilder->setMetadata($reservationData[ReservationInterface::METADATA]);
+
+        $this->assertSame($this->reservation, $reservationBuilder->build());
+    }
```

Comment:
Please add Test covering Negative cases, when validation not passed and we need to throw a ValidationException

Reply:
I may be wrong but in ReservationBuilder the `throw new ValidationException($validationResult);` doesn't allow me to mock the `ValidationException` object so I will incur in issue #81.

Let me know if I can use a workaround to avoid direct exception instantiation or we have to wait for issue #81 to be fixed.

Reply:
you can use in your Unit Test
```php
$this->expectException(ValidationException::class);
```

That will make sure that Exception has been thrown.

In addition to the expectException() method the expectExceptionCode(), expectExceptionMessage(), and expectExceptionMessageRegExp() methods exist to set up expectations for exceptions raised by the code under test.

Alternatively, you can use the @expectedException, @expectedExceptionCode, @expectedExceptionMessage, and @expectedExceptionMessageRegExp annotations to set up expectations for exceptions raised by the code under test.

Reply:
I tried but the `ValidationException` is not thrown because when I instance it with `throw new ValidationException($validationResult);` in `ReservationBuilder::build()` I incur in issue #81 ang get the following error:
```
Magento\Inventory\Test\Unit\Model\ReservationBuilderTest::testThrowValidationException with data set #0 (array(1, 'somesku', 11, 'some meta data'), array(1, 'somesku', 11, 'some meta data'))
TypeError: Argument 1 passed to Magento\Framework\Exception\AbstractAggregateException::__construct() must be an instance of Magento\Framework\Phrase, null given, called in /app/magento2/lib/internal/Magento/Framework/Validation/ValidationException.php on line 27

/app/magento2/lib/internal/Magento/Framework/Exception/AbstractAggregateException.php:44
/app/magento2/lib/internal/Magento/Framework/Validation/ValidationException.php:27
/app/magento2/app/code/Magento/Inventory/Model/ReservationBuilder.php:113
/app/magento2/app/code/Magento/Inventory/Test/Unit/Model/ReservationBuilderTest.php:139
```

That's why I tried to mock the `ValidationException` but creating it via Object Manager and then throwing it doesn't seem to work.